File: neo4j/src/main/scala/akka/stream/alpakka/neo4j/scaladsl/ReactiveNeo4j.scala
/*
 * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
 */
package akka.stream.alpakka.neo4j.scaladsl

import akka.stream.alpakka.neo4j.internal.codec.{Neo4jDecoder, Neo4jEncoder}
import akka.stream.alpakka.neo4j.internal.stage.{Neo4jFiniteSourceStage, Neo4jFlowStage}
import akka.stream.alpakka.neo4j.internal.{CypherMarshaller, CypherUnmarshaller, Neo4jCapabilities}
import akka.stream.scaladsl.{Flow, Source}
import akka.{Done, NotUsed}
import org.neo4j.driver.v1.{Driver, Statement, Values}

import scala.concurrent.Future

class ReactiveNeo4j(driver: Driver) extends Neo4jCapabilities(driver) {

  /**
   * Shapeless powered flow.
   * @param label given to nodes
   * @param encoder shapeless encoder
   */
  def flow[A](label: Symbol)(implicit encoder: Neo4jEncoder[A]): Flow[A, A, NotUsed] =
    Flow.fromGraph(new Neo4jFlowStage[A](driver, new ShapelessCypherMarshaller(label.name, encoder)))

  /**
   * Flow with hand crafted marshaller
   * @param writeHelper
   * @param label given to node
   */
  def flow[A](writeHelper: CypherMarshaller[A], label: Symbol): Flow[A, A, NotUsed] =
    Flow.fromGraph(new Neo4jFlowStage[A](driver, writeHelper))

  /**
   * Shapeless powered source.
   * @param decoder shapeless decoder
   */
  def source[A](statement: Statement)(implicit decoder: Neo4jDecoder[A]): Source[A, Future[Done]] =
    Source.fromGraph(new Neo4jFiniteSourceStage(driver, new ShapelessCypherUnmarshaller[A](decoder), statement))

  /**
   * Shapeless powered source.
   * @param decoder shapeless decoder
   */
  def source[A](statement: String, args: String*)(implicit decoder: Neo4jDecoder[A]): Source[A, Future[Done]] =
    source(new Statement(statement, Values.parameters(args: _*)))(decoder)

  /**
   * Source with hand crafted unmarshaller
   * @param readedHelper
   */
  def source[A](readedHelper: CypherUnmarshaller[A], statement: Statement): Source[A, Future[Done]] =
    Source.fromGraph(new Neo4jFiniteSourceStage(driver, readedHelper, statement))

  /**
   * Source with hand crafted unmarshaller
   * @param readedHelper
   */
  def source[A](readedHelper: CypherUnmarshaller[A], statement: String, args: String*): Source[A, Future[Done]] =
    source(readedHelper, new Statement(statement, Values.parameters(args: _*)))

}

File: neo4j/src/main/java/akka/stream/alpakka/neo4j/javadsl/ReactiveNeo4j.java
/*
 * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
 */
package akka.stream.alpakka.neo4j.javadsl;

import akka.Done;
import akka.NotUsed;
import akka.stream.alpakka.neo4j.internal.CypherMarshaller;
import akka.stream.alpakka.neo4j.internal.CypherUnmarshaller;
import akka.stream.alpakka.neo4j.internal.Neo4jCapabilities;
import akka.stream.alpakka.neo4j.internal.stage.Neo4jFiniteSourceStage;
import akka.stream.alpakka.neo4j.internal.stage.Neo4jFlowStage;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Source;
import org.neo4j.driver.v1.Driver;
import org.neo4j.driver.v1.Statement;
import org.neo4j.driver.v1.Values;
import scala.concurrent.Future;


public class ReactiveNeo4j extends Neo4jCapabilities {


    public ReactiveNeo4j(Driver driver) {
        super(driver);
    }

    public <A> Source<A, Future<Done>> source(CypherUnmarshaller<A> readHelper, Statement statement) {
        return Source.fromGraph(new Neo4jFiniteSourceStage<A>(driver(), readHelper, statement));
    }

    public <A> Source<A, Future<Done>> source(CypherUnmarshaller<A> readHelper, String query, Object ... args) {
        return source(readHelper, new Statement(query, Values.parameters(args)));
    }


    public <A> Flow<A,A,NotUsed> flow(String label, CypherMarshaller<A> writerHelper) {
        return Flow.fromGraph(new Neo4jFlowStage<A>(driver(), writerHelper));
    }
}
